src/analyses/base.ml

```diff
@@ -87,14 +87,50 @@ struct
 
   let is_privglob v = GobConfig.get_bool "annotation.int.privglobs" && v.vglob
 
-  let project_val p_opt value is_glob =
-    match GobConfig.get_bool "annotation.int.enabled", is_glob, p_opt with
-    | true, true, _ -> VD.project PU.max_int_precision value
-    | true, false, Some p -> VD.project p value
-    | _ -> value
+  (*This is a bit of a hack to be able to change array domains if a pointer to an array is given as an argument*)
+  (*We have to prevent different domains to be used at the same time for the same array*)
+  (*After a function call, the domain has to be the same as before and we can not depend on the pointers staying the same*)
+  (*-> we determine the arrays a pointer can point to once at the beginning of a function*)
+  (*There surely is a better way, because this means that often the wrong one gets chosen*)
```

Comment:
This looks very suspicious, especially because it ignores contexts and very likely breaks incrementally (at minimum it should be marshaled or cleared for every analysis? not sure what the scope of this information is supposed to be).

How is this different from the existing int domain projections? Pointers to int variables exist as well.

Reply:
The array annotations are scoped per array/type and not the function like the int domain. 
I think this is more useful for them because for them their individual usage is more important and opposed to integers there are no problems if two different arrays have different domains.

If we then do want to change them for some function only, my best solution was to do this for the pointer parameters, as array arguments do not exist. As a result of time restraints, this is not really used yet. 

I also dislike that this makes the function somewhat stateful.
The scope of this should be the analysis of a single program. Would it be enough for that to reset it at the same place as all the lazy  maps?
I did not think about the incremental analysis and lack the knowledge to see how this could be accommodated.

Reply:
> Would it be enough for that to reset it at the same place as all the lazy maps?

I think for now that probably is the best option here. Compatibility with incremental is something that is probably out of scope for your thesis, and we can worry about it later.

Thinking of it, maybe we simply make a check that `annotation.array` and `incremental.load` are not activated at the same time for now?

Reply:
> Would it be enough for that to reset it at the same place as all the lazy maps?

Doing the resetting in the `Spec`'s `init` and `finalize` should be enough, not to pull this table into an even more global scope.

> Thinking of it, maybe we simply make a check that `annotation.array` and `incremental.load` are not activated at the same time for now?

I think it might even be fine: if a function doesn't get incrementally reanalyzed, then this won't be looked up again anyway, and if it does, then it gets added to the map again. Resetting it in `init` would also give this behavior in server mode.

Reply:
I just reset it in finalize. I think normally that should be enough to do it at one point,  not before and after. Does it need both for incremental?

Reply:
That should be enough for this indeed.